### PR TITLE
Merging EI maintenance branch to master branch

### PR DESCRIPTION
**Pull request by zburswa**
Added the changes from EI maintenance branch to master branch

### Description of the Change
Added the commits from EI backend maintenance branch to master branch.
These are the commits which were added:
https://github.com/eiffel-community/eiffel-intelligence/commit/bc766553f7b08dc84254b41facf420deb32e0c92
https://github.com/eiffel-community/eiffel-intelligence/commit/0637e9fc393a8f6a84870c4a9723625594a54907
https://github.com/eiffel-community/eiffel-intelligence/commit/5461175978de54dcc8ea5dd0a564ee838fd21460
https://github.com/eiffel-community/eiffel-intelligence/commit/6d1d5dd14fcf84b6d723714f15dc74642fc6c450
https://github.com/eiffel-community/eiffel-intelligence/commit/ce574ffe79d9ca11fd2b95d9673c2bae0e7e230a
https://github.com/eiffel-community/eiffel-intelligence/commit/36f94614a425cf10fb31a8283c4ac7e6744a20d0

EI- frontend master branch is even with maintenance branch.

### CODE REVIEW

**Comment by e-pettersson-ericsson** [comment]
Remove the DURACI reference in this PR, it does not belong on GitHub.

**Comment by tobiasake** [comment]
Looks ok, but there is a lot of MongoDB handling code changes.
We need to test so the MongoDb/RabbitMq restoring connections still works which was fixed in EI 3.1.0 version release, before making a new EI release.
This can be tested after all 2.0 maintenance branch fixes has been added to master branch.

**Comment by zburswa** [comment]
> Looks ok, but there is a lot of MongoDB handling code changes.
> We need to test so the MongoDb/RabbitMq restoring connections still works which was fixed in EI 3.1.0 version release, before making a new EI release.
> This can be tested after all 2.0 maintenance branch fixes has been added to master branch.

ok Tobias.

File: .travis.yml [hunk 1]
```diff
@@ -43,6 +43,6 @@ jobs:
       name: report_coverage
       # Generating test coverage report and publishing to Codacy
       script:
-        - mvn jacoco:prepare-agent test jacoco:report -B
-        - curl -Ls -o codacy-coverage-reporter-assembly.jar https://github.com/codacy/codacy-coverage-reporter/releases/download/6.0.2/codacy-coverage-reporter-6.0.2-assembly.jar
-        - java -jar codacy-coverage-reporter-assembly.jar report -l Java -r target/site/jacoco/jacoco.xml
+        - mvn cobertura:cobertura -B
+        - curl -LSs $(curl -LSs https://api.github.com/repos/codacy/codacy-coverage-reporter/releases/latest | jq -r '.assets | map({content_type, browser_download_url} | select(.content_type | contains("application/octet-stream"))) | .[0].browser_download_url') -o codacy-coverage-reporter-assembly.jar
```
Comment (SantoshNC68) [inline .travis.yml:47]: Is this change really required? Why was it modified in the master. 
@e-pettersson-ericsson / @a-breid-ericsson  can you comment if this change is required in merging the maintenance branch to master?

File: src/functionaltests/java/com/ericsson/ei/notifications/ttl/TestTTLSteps.java [hunk 2]
```diff
@@ -108,25 +109,30 @@ public void create_subscription_object() throws IOException, JSONException {
...
+                "utf-8");
 
         // replace with port of running mock server
-        subscriptionStr = subscriptionStr.replaceAll("\\{port\\}", String.valueOf(clientAndServer.getPort()));
+        subscriptionStr = subscriptionStr.replaceAll("\\{port\\}",
+                String.valueOf(clientAndServer.getPort()));
 
         subscriptionObject = new ObjectMapper().readTree(subscriptionStr);
-        assertEquals(false, subscriptionObject.get("notificationMeta").toString().contains("{port}"));
+        assertEquals(false,
```
Comment (SantoshNC68) [inline src/functionaltests/java/com/ericsson/ei/notifications/ttl/TestTTLSteps.java:120]: do not modify if this is not a change.

File: src/functionaltests/java/com/ericsson/ei/notifications/ttl/TestTTLSteps.java [hunk 3]
```diff
@@ -108,25 +109,30 @@ public void create_subscription_object() throws IOException, JSONException {
         LOGGER.debug("Starting scenario @TestNotificationRetries.");
         mongoDBHandler.dropCollection(database, failedNotificationCollection);
 
-        String subscriptionStr = FileUtils.readFileToString(new File(SUBSCRIPTION_FILE_PATH), "utf-8");
+        String subscriptionStr = FileUtils.readFileToString(new File(SUBSCRIPTION_FILE_PATH),
+                "utf-8");
 
         // replace with port of running mock server
-        subscriptionStr = subscriptionStr.replaceAll("\\{port\\}", String.valueOf(clientAndServer.getPort()));
+        subscriptionStr = subscriptionStr.replaceAll("\\{port\\}",
```
Comment (SantoshNC68) [inline src/functionaltests/java/com/ericsson/ei/notifications/ttl/TestTTLSteps.java:116]: do not modify if this is not a change.

File: .travis.yml (hunk 1 above)
Comment (e-pettersson-ericsson) [reply .travis.yml:47]: travis.yml 14 months ago:
https://github.com/eiffel-community/eiffel-intelligence/blame/2.0.0-maintenance/.travis.yml

travis.yml 12 months ago, on master: 
https://github.com/eiffel-community/eiffel-intelligence/blame/master/.travis.yml

I guess taking the latest changes would be the safest, but Codacy does not seem to be enabled at all anymore, so a separate issue should be created and investigated how to enable it again.

File: src/functionaltests/java/com/ericsson/ei/rabbitmq/connection/RabbitMQTestConnectionSteps.java [hunk 4]
```diff
@@ -56,10 +56,11 @@
     private String rabbitMQPort;
 
     private static final Logger LOGGER = LoggerFactory.getLogger(RabbitMQTestConnectionSteps.class);
-    private static final String EIFFEL_EVENTS = "src/functionaltests/resources/eiffel_events_for_thread_testing.json";
+    private static final String EIFFEL_EVENTS = "src/functionaltests/resources/eiffel_events_for_test.json";
```
Comment (SantoshNC68) [inline src/functionaltests/java/com/ericsson/ei/rabbitmq/connection/RabbitMQTestConnectionSteps.java:59]: Have you checked the differences between these files? If it is just a naming thing or any changes in json itself?

Comment (zburswa) [reply src/functionaltests/java/com/ericsson/ei/rabbitmq/connection/RabbitMQTestConnectionSteps.java:59]: > Have you checked the differences between these files? If it is just a naming thing or any changes in json itself?

Yes. Instead of testing with 4 events, here we have tested with only 1 event. So, only 1 event is being declared.

**Review by SantoshNC68 — APPROVED** [review]

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java [hunk 5]
```diff
@@ -130,7 +146,7 @@ public String getEventId(RulesObject rulesObject, String event) {
             try {
                 JsonNode document = mapper.readValue(mapStr, JsonNode.class);
                 JsonNode value = document.get(listPropertyName);
-                list = new ObjectMapper().readValue(value.traverse(), new TypeReference<ArrayList<String>>(){});
+                list = new ObjectMapper().readValue(value.traverse(), new TypeReference<ArrayList<String>>() {});
             } catch (Exception e) {
                 LOGGER.info("Failed to deserialize event object list.", e);
```
Comment (sudharshan-bandaru) [inline src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:151]: Change it info to error

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java [hunk 6]
```diff
@@ -157,5 +173,4 @@ public boolean isEventInEventObjectMap(String eventId) {
         return !documents.isEmpty();
```
Comment (sudharshan-bandaru) [inline src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:173]: if possible try to change this line of code

File: src/main/java/com/ericsson/ei/mongo/MongoDBHandler.java [hunk 7]
```diff
@@ -218,13 +226,17 @@ public boolean dropDocument(String dataBaseName, String collectionName, MongoQue
      * @param fieldName      for index creation field
      * @param ttlValue       seconds
      */
```
Comment (sudharshan-bandaru) [inline src/main/java/com/ericsson/ei/mongo/MongoDBHandler.java:228]: Add the javadoc about MongoDBConnectionException

File: src/main/java/com/ericsson/ei/rules/IRuleTestService.java [hunk 8]
```diff
@@ -3,6 +3,8 @@
 import java.io.IOException;
 
 import com.ericsson.ei.exception.InvalidRulesException;
+import com.ericsson.ei.exception.MongoDBConnectionException;
```
Comment (sudharshan-bandaru) [inline src/main/java/com/ericsson/ei/rules/IRuleTestService.java:6]: Remove this import statement

**Review by sudharshan-bandaru — APPROVED** [review]
